arbor/backends/event_stream_base.hpp

```diff
@@ -62,24 +59,31 @@ struct event_stream_base {
         index_ = 0;
     }
 
-    // Construct a mapping of mech_id to a stream s.t. streams are partitioned into
-    // time step buckets by `ev_span`
+  protected:
+    // backend specific initializations
+    virtual void init() = 0;
+};
+
+struct spike_event_stream_base : event_stream_base<deliverable_event> {
     template<typename EventStream>
-    static std::enable_if_t<std::is_base_of_v<event_stream_base, EventStream>>
-    multi_event_stream(const event_lane_subrange& lanes,
-                       const std::vector<target_handle>& handles,
-                       const std::vector<std::size_t>& divs,
-                       const timestep_range& steps,
-                       std::unordered_map<unsigned, EventStream>& streams) {
+    friend void initialize(const event_lane_subrange& lanes,
+                           const std::vector<target_handle>& handles,
+                           const std::vector<std::size_t>& divs,
+                           const timestep_range& steps,
+                           std::unordered_map<unsigned, EventStream>& streams) {
         arb_assert(lanes.size() < divs.size());
 
+        // reset streams and allocate sufficient space for temporaries
         auto n_steps = steps.size();
-        std::unordered_map<unsigned, std::vector<std::size_t>> dt_sizes;
         for (auto& [k, v]: streams) {
             v.clear();
-            dt_sizes[k].resize(n_steps, 0);
+            v.spike_counter_.clear();
+            v.spike_counter_.resize(steps.size(), 0);
+            v.spikes_.clear();
+            v.spikes_.reserve(v.ev_data_.capacity());
```

Comment:
this is intentional, as `ev_data_` has been cleared before. I should probably make a comment. Otherwise, I can reorder the operations, so that `v.clear()` is called afterwards.

Reply:
I think that would be confusing, just make a note.